### DISCUSSION TITLE
Code optimization

### DISCUSSION THREAD

**Discussion by Dev-ALPM** [imixs/imixs-workflow#839] [Ideas]
Good morning,
looking through the code, I saw:

- using the Vector class which is synchronized. I think you can change it to the ArrayList class in most cases.

- String concatenation can be improved using a StringBuilder

- Unnecessary type conversions

I can take care of it and measure the performance gains

**Comment by rsoika** [comment]
Hi  @Dev-ALPM ,

yes, Imixs-Workflow has a very long history ;-) The Vector class is still a relict from old days in some places. Yes of course this could and should be replaced with `ArrayList`. But similar to the String concatenation vs `StringBuilder` I would not expect measurable performance improvements.

I don't think it would be worth the effort here now. But of course, these points should be taken into account for future changes or bug fixes.

Anyway, I wanted to ask you which IDE you use and if you have an idea to install better style checkers? Code Checking and formatting is still not perfect

**Reply by Dev-ALPM** [reply]
I use Netbeans.

For code style checking you can use the maven plugin:
[https://maven.apache.org/plugins/maven-checkstyle-plugin/index.html](https://maven.apache.org/plugins/maven-checkstyle-plugin/index.html)

For automatic formatting before compiling the maven plugin:
[https://github.com/revelc/formatter-maven-plugin](https://github.com/revelc/formatter-maven-plugin)
